drivers/frequency/adf5902/adf5902.c

```diff
@@ -0,0 +1,946 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include <malloc.h>
+#include <math.h>
+#include "adf5902.h"
+#include "error.h"
+#include "delay.h"
+
+/******************************************************************************/
+/************************** Functions Implementation **************************/
+/******************************************************************************/
+
+/**
+ * @brief Writes 4 bytes of data to ADF5902.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param data - Data value to write.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int32_t adf5902_write(struct adf5902_dev *dev, uint8_t reg_addr, uint32_t data)
+{
+	uint8_t buff[ADF5902_BUFF_SIZE_BYTES];
+
+	data = data | reg_addr;
+
+	buff[0] = data >> 24;
+	buff[1] = data >> 16;
+	buff[2] = data >> 8;
+	buff[3] = data;
+
+	return spi_write_and_read(dev->spi_desc, buff, ADF5902_BUFF_SIZE_BYTES);
+}
+
+/**
+ * @brief Readback data from ADF5902.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param data - Store the read data.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int32_t adf5902_readback(struct adf5902_dev *dev, uint8_t reg_addr,
+			 uint32_t *data)
+{
+	int32_t ret;
+	uint32_t i;
+	uint8_t buff[ADF5902_BUFF_SIZE_BYTES];
+
+	*data = *data | reg_addr;
+
+	buff[0] = *data >> 24;
+	buff[1] = *data >> 16;
+	buff[2] = *data >> 8;
+	buff[3] = *data;
+
+	/* Write command */
+	ret = spi_write_and_read(dev->spi_desc, buff, ADF5902_BUFF_SIZE_BYTES);
+	if (ret != SUCCESS)
+		return ret;
+
+	/* Send dummy value to get the data on DOUT */
+	buff[0] = ADF5902_SPI_DUMMY_DATA;
+	buff[1] = ADF5902_SPI_DUMMY_DATA;
+	buff[2] = ADF5902_SPI_DUMMY_DATA;
+	buff[3] = ADF5902_SPI_DUMMY_DATA;
+
+	ret = spi_write_and_read(dev->spi_desc, buff, ADF5902_BUFF_SIZE_BYTES);
+	if (ret != SUCCESS)
+		return ret;
+
+	/* Concatenate received data Bytes */
+	for (i = 0; i < ADF5902_BUFF_SIZE_BYTES; i++)
+		*data = (*data << 8) | buff[i];
+
+	return ret;
+}
+
+/**
+ * @brief Compute ADF4350 RF VCO frequency parameters.
+ * @param dev - The device structure.
+ * @return Returns 0 in case of success or negative error code.
+ */
+static int32_t adf5902_vco_freq_param(struct adf5902_dev *dev)
+{
+
+	dev->f_pfd = (dev->ref_in * (1 + dev->ref_doubler_en) /
+		      (dev->ref_div_factor *(1 + dev->ref_div2_en)));
+
+	if (dev->f_pfd > ADF5902_MAX_FREQ_PFD)
```

Comment:
Can't we compute that all internally? Try to maximize PFD

Reply:
This is how documentation suggests the PFD computation based on the input parameters:
https://www.analog.com/media/en/technical-documentation/data-sheets/ADF5902.pdf
(Page 11 of 39)
![image](https://user-images.githubusercontent.com/26061529/108845246-d6678080-75e5-11eb-8fc4-e9ad116e762c.png)

What do you mean by "Try to maximize PFD"?

Reply:
For example, in your formula, the f_pfd depends on ref_div_factor. You should calculate the ref_div_factor (instead of providing it in the initialization structure) so the f_pfd value will be the maximum allowed one (still lower than ADF5902_MAX_FREQ_PFD).

Reply:
I understand, thanks! I was using as reference the ADF5902 Evaluations Software from analog.com, where all these parameters were configured manually by the user (including ref_div_factor). I will compute them inside the driver.